src/bin/cql-stress-cassandra-stress/settings/param/multi_param.rs

```diff
@@ -0,0 +1,266 @@
+/// replication(foo=bar,factor=3,key=value)
+///
+/// The multiparameter will delegate parsing of `factor=3` part to its predefined subparameter.
+/// `foo=bar` and `key=value` will be stored in the map of arbitrary parameters.
+pub struct MultiParam {
+    prefix: &'static str,
+    // Pre-defined parameters.
+    // User can access them via their corresponding handles.
+    subparams: Vec<ParamCell>,
+    desc: &'static str,
+    required: bool,
+    // Arbitrary parameters of the `key=value` form.
+    arbitrary_params: Option<HashMap<String, String>>,
+    supplied_by_user: bool,
+    satisfied: bool,
+}
+
+impl MultiParam {
+    pub fn new(
+        prefix: &'static str,
+        subparams: Vec<ParamCell>,
+        desc: &'static str,
+        required: bool,
+        accept_arbitrary: bool,
+    ) -> Self {
+        Self {
+            prefix,
+            subparams,
+            desc,
+            required,
+            arbitrary_params: if accept_arbitrary {
+                Some(HashMap::new())
+            } else {
+                None
+            },
+            supplied_by_user: false,
+            satisfied: false,
+        }
+    }
+
+    fn accepts_arbitrary(&self) -> bool {
+        self.arbitrary_params.is_some()
+    }
+
+    /// Retrieves arbitrary subparameters (if parsed successfully) and consumes the parameter.
+    pub fn get_arbitrary(self) -> Option<HashMap<String, String>> {
+        if !self.satisfied {
+            return None;
+        }
+        self.arbitrary_params
+    }
+
+    fn try_parse_predefined(&self, arg: &str) -> ParamMatchResult {
+        for param in self.subparams.iter() {
+            let mut borrowed = param.borrow_mut();
+            match borrowed.try_match(arg) {
+                ParamMatchResult::NoMatch => (),
+                ParamMatchResult::Error(e) => return ParamMatchResult::Error(e),
+                ParamMatchResult::Match => match borrowed.parse(arg) {
+                    Ok(()) => return ParamMatchResult::Match,
+                    Err(e) => return ParamMatchResult::Error(e),
+                },
+            }
+        }
+
+        ParamMatchResult::NoMatch
+    }
+
+    fn try_parse_arbitrary(&mut self, arg: &str) -> Result<()> {
+        // Ensure that argument matches pattern "key=value"
+        anyhow::ensure!(
+            ARBITRARY_PARAM.is_match(arg),
+            "Invalid {} specification: {}",
+            self.prefix,
+            arg
+        );
+
+        let split = arg.split('=').collect::<Vec<&str>>();
+        anyhow::ensure!(split.len() == 2, "Invalid arbitrary parameter: {}", arg);
+
+        let map = self.arbitrary_params.as_mut().unwrap();
+        let key = split[0].to_string();
+        let val = split[1].to_string();
+        anyhow::ensure!(
+            !map.contains_key(&key),
+            "{} suboption has been specified more than once",
+            key
+        );
+        map.insert(key, val);
```

Comment:
> Is it a case that every multiparameter statically knows whether it accepts arbitrary parameters or not?

Yes, the `Param`s API objects are created internally, and are predefined (based on the parameters we support). So we can for sure program the so called "arbitrary parameters acceptance" using the Rust types (instead of boolean flag).

> My approach to the typesafe pattern here:

Thank you for presenting your approach. I understand the pattern and see the benefits. I have only one question. Let's take the `MultiParam::parse` function:

```rust
fn parse(&mut self, arg: &str) -> Result<()> {
        self.supplied_by_user = true;
        let arg_val = &arg[self.prefix.len()..];

        // Remove wrapping paranthesis.
        let arg_val = &arg_val[1..arg_val.len() - 1];

        // Iterate over comma-delimited sub-parameters.
        for subparam in arg_val.split(',') {
            // Check if the argument matches on of the predefined subparameters.
            match self.try_parse_predefined(subparam) {
                ParamMatchResult::Error(e) => return Err(e),
                ParamMatchResult::Match => continue,
                _ => (),
            }

            // If the argument didn't match any of the prefefined sub-parameters,
            // try to parse it as an arbitrary parameter (if applicable).
            anyhow::ensure!(
                self.accepts_arbitrary(),
                "Cannot accept parameter {}. {} command/option doesn't accept arbitrary parameters.",
                subparam,
                self.prefix
            );

            self.try_parse_arbitrary(subparam)?;
        }

        Ok(())
  }
``` 

From what I understood, in the snippet you provided, the `try_parse_arbitrary` is not implemented for `MultiParam<RejectsArbitratyParams>`. So now, we would technically need to implement `parse` method for both of the types separately (since it calls `try_parse_arbitrary` in one case, and doesn't call it in the other case).

In this particular case, the common `parse` logic is very simple, so there would be no problem copying this part of the code twice (or even moving it to some function implemented by `impl<T: ArbitraryParamAcceptance> MultiParam<T>`, and then calling the common logic from the actual implementations for the specific types). But what if the common logic would be much more complex?

But this is not the only solution. The other solution I see is to add `try_parse_arbitrary` method to the `ArbitraryParamsAcceptance` trait. In the case of `AcceptsArbitraryParams`, we would have the current implementation of `try_parse_arbitrary`, and in the case of `RejectsArbitraryParams` we would do nothing (nop), or maybe return the error (if for some reason the programmer didn't check if `accepts_arbitrary` before calling `try_parse_arbitrary).

What do you think about it?

Reply:
I've left this part unimplemented on purpose, for you to consider various possibilities :)
Your conclusions are in line with mine. Duplicating the logic is unacceptable; therefore, I'd go for the solution with extracting these lines:
```rust
    // If the argument didn't match any of the prefefined sub-parameters,
    // try to parse it as an arbitrary parameter (if applicable).
    anyhow::ensure!(
        self.accepts_arbitrary(),
        "Cannot accept parameter {}. {} command/option doesn't accept arbitrary parameters.",
        subparam,
        self.prefix
    );

    self.try_parse_arbitrary(subparam)?;
```
into a function whose implementation depends on `ArbitraryParamsAcceptance` type.
The implementation for `RejectArbitraryParams` would return the error:
```rust
    anyhow::anyhow!("Cannot accept parameter {}. {} command/option doesn't accept arbitrary parameters.",
                subparam,
                self.prefix
    )
```
and the one for `AcceptArbitraryArguments` would just call `try_parse_arbitrary()`.

The idea is, once again, that we want to make it **IMPOSSIBLE** to make an error, so we want the programmer to not have to call `accepts_arbitrary()` anytime here.